File: ports/go/src/gui/painting/bcsbrush.go
package painting

import (
	"sync"
	"github.com/robertpelloni/btk-go/ports/go/src/core/kernel"
)

type BrushStyle int
const (
	NoBrush BrushStyle = iota
	SolidPattern
	Dense1Pattern
	Dense2Pattern
	Dense3Pattern
	Dense4Pattern
	Dense5Pattern
	Dense6Pattern
	Dense7Pattern
	HorPattern
	VerPattern
	CrossPattern
	BDiagPattern
	FDiagPattern
	DiagCrossPattern
	LinearGradientPattern
	RadialGradientPattern
	ConicalGradientPattern
	TexturePattern
)

// BcsBrush maps to QBrush
type BcsBrush struct {
	*kernel.BcsObject
	style BrushStyle
	color string // Using string for color hex code simplistically
	mu    sync.RWMutex
}

func NewBcsBrush() *BcsBrush {
	return &BcsBrush{
		BcsObject: kernel.NewBcsObject(nil),
		style:     NoBrush,
		color:     "#000000",
	}
}

func NewBcsBrushWithColor(color string, style BrushStyle) *BcsBrush {
	return &BcsBrush{
		BcsObject: kernel.NewBcsObject(nil),
		style:     style,
		color:     color,
	}
}

func (b *BcsBrush) SetStyle(s BrushStyle) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.style = s
}

func (b *BcsBrush) Style() BrushStyle {
	b.mu.RLock()
	defer b.mu.RUnlock()
	return b.style
}

func (b *BcsBrush) SetColor(color string) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.color = color
}

func (b *BcsBrush) Color() string {
	b.mu.RLock()
	defer b.mu.RUnlock()
	return b.color
}

File: ports/rust/src/gui/painting/bcs_brush.rs
use std::sync::{Arc, Mutex};
use crate::core::kernel::bcs_object::BcsObject;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BrushStyle {
    NoBrush = 0,
    SolidPattern,
    Dense1Pattern,
    Dense2Pattern,
    Dense3Pattern,
    Dense4Pattern,
    Dense5Pattern,
    Dense6Pattern,
    Dense7Pattern,
    HorPattern,
    VerPattern,
    CrossPattern,
    BDiagPattern,
    FDiagPattern,
    DiagCrossPattern,
    LinearGradientPattern,
    RadialGradientPattern,
    ConicalGradientPattern,
    TexturePattern,
}

pub struct BcsBrush {
    pub base: Arc<BcsObject>,
    style: Mutex<BrushStyle>,
    color: Mutex<String>,
}

impl BcsBrush {
    pub fn new() -> Arc<Self> {
        Arc::new(Self {
            base: BcsObject::new(),
            style: Mutex::new(BrushStyle::NoBrush),
            color: Mutex::new(String::from("#000000")),
        })
    }

    pub fn new_with_color_style(color: &str, style: BrushStyle) -> Arc<Self> {
        Arc::new(Self {
            base: BcsObject::new(),
            style: Mutex::new(style),
            color: Mutex::new(color.to_string()),
        })
    }

    pub fn set_style(&self, style: BrushStyle) {
        *self.style.lock().unwrap() = style;
    }

    pub fn style(&self) -> BrushStyle {
        *self.style.lock().unwrap()
    }

    pub fn set_color(&self, color: &str) {
        *self.color.lock().unwrap() = color.to_string();
    }

    pub fn color(&self) -> String {
        self.color.lock().unwrap().clone()
    }
}
